### ISSUE TITLE
Math behind dsstne

### ISSUE THREAD

**Issue by IlyaTrofimov** [amznlabs/amazon-dsstne#51]
Hi guys,

Thank you for releasing your project to open-source.
Can you please explain a math behing dsstne?

Is it a collaborative filtering with explicit/implicit feedback?
You don't use ratings from MovieLens, so I assume that implict...
A link to a paper/doc will be very userfull.

**Comment by scottlegrand** [comment]
DSSTNE is a general Neural Network framework with an emphasis on highly sparse data (data density of 1% or less).  The example provided can beat TensorFlow (which appears to rely on CuSparse) by a huge margin, even on a single GPU.

As for the math, it's forward and backpropagation.  As for recommendations, the example provided constructs an autoencoder-like network of movie viewing indicators for training.  If one then sorts the output of this autoencoder on a held back test set of data, filtering out previously viewed movies, one gets its recommendations. 

That said, there's no specific need to treat training as an autoencoder.  Any set of input data could be used to predict any set of output data.

**Comment by IlyaTrofimov** [comment]
Autoencoder! I get the idea.

When input and output data were generated from one dataset (ml20m-all)

`generateNetCDF -d gl_input -i ml20m-all -o gl_input.nc -f features_input -s samples_input -c
generateNetCDF -d gl_output -i ml20m-all -o gl_output.nc -f features_output -s samples_input -c`

network works as an autoencoder.
So by generating input data from my X dataset, output data from my Y dataset - 
network will do regression.
I am right?

**Comment by scottlegrand** [comment]
Close, by switching the Y data set and doing nothing else, you'd be regressing the inputs to 1 or 0

To do full regression, you'd also have to change the loss function on the output to L2 or L1 and use analog sparse data (supported) to provide something beyond 1 or 0 as output targets, but yes.

Do you need example code for the latter case?

**Comment by IlyaTrofimov** [comment]
yes please

**Comment by scottlegrand** [comment]
Cool, I've attached a short C++ program and the command to build the dataset with output data set to the ratings of each movie from movieLens.  A network to predict ratings would require some additional design and hyperparameter tuning, but here's how to get the data in the right form to begin addressing that.

[issue51.zip](https://github.com/amznlabs/amazon-dsstne/files/378280/issue51.zip)

**Comment by IlyaTrofimov** [comment]
Thank you, Scott. I've succesfully transformed ratings.csv from ml-20m.zip to a proper format using your tool.
Now I'm going to adapt it to my case.
